### PR TITLE
Enable Signing in Repos that Override Build Actions

### LINKED ISSUE #4064: Enable signing for arcade repos in the VMR build
Enable signing for arcade repos in the VMR build. This may be as simple as ensuring that -sign gets passed? This may not be sufficient though, as signing may have been disabled under UB modes in Arcade at some point. Investigate and fix.

It should be possible to do this work on Windows VMR legs without the other SignTool work being completed.

### PR DESCRIPTION
**Pull request by ellahathaway**
Closes https://github.com/dotnet/source-build/issues/4064

Enables signing for the following repos which were not passing `-sign` under UB modes:
- aspnetcore
- fsharp
- roslyn

### CODE REVIEW

File: src/SourceBuild/content/repo-projects/fsharp.proj
Comment (ViktorHofer) [inline src/SourceBuild/content/repo-projects/fsharp.proj:11]: Why did you add the `DotNetBuildSourceOnly` condition only to aspnetcore.proj?

File: src/SourceBuild/content/repo-projects/nuget-client.proj
Comment (ViktorHofer) [inline src/SourceBuild/content/repo-projects/nuget-client.proj:14]: Are you sure nuget supports this? The comment in L7 suggests otherwise.

Comment (ellahathaway) [reply src/SourceBuild/content/repo-projects/nuget-client.proj]: I just realized this... this will get handled by https://github.com/dotnet/source-build/issues/4065

File: src/SourceBuild/content/repo-projects/fsharp.proj
Comment (ellahathaway) [reply src/SourceBuild/content/repo-projects/fsharp.proj]: This was missed on my part. Fixed in https://github.com/dotnet/sdk/pull/44855/commits/4fed5a3acbedb056a3a1c1a82976edf5f7d09506

**Review by ViktorHofer — APPROVED** [review]

**Comment by ellahathaway** [comment]
Looks like the `sign` arg is not valid in fsharp, and /p:Sign is not recognized either. I'll have to find a workaround for this before this PR is ready to be merged,

**Comment by ellahathaway** [comment]
Signing aspnetcore is failing because there are no ItemsToSign defined in Signing.props: https://github.com/dotnet/aspnetcore/blob/main/eng/Signing.props#L7-L9

Signing aspnetcore artifacts in the VMR needs to be enabled, following up in https://github.com/dotnet/aspnetcore/issues/58445#issuecomment-2476858792. In the meantime, I'm going to draft this PR.

**Comment by ViktorHofer** [comment]
> Looks like the sign arg is not valid in fsharp, but I tested passing /p:Sign=true to fsharp and that worked to invoke signing. Updated in https://github.com/dotnet/sdk/commit/9ef5b628515022996c424841459000de59d3de2c

That's weird. Can you please file an issue in fsharp to track allowing the `-sign` switch?

**Comment by ellahathaway** [comment]
> > Looks like the sign arg is not valid in fsharp, but I tested passing /p:Sign=true to fsharp and that worked to invoke signing. Updated in [9ef5b62](https://github.com/dotnet/sdk/commit/9ef5b628515022996c424841459000de59d3de2c)
> 
> That's weird. Can you please file an issue in fsharp to track allowing the `-sign` switch?

Yes, filed https://github.com/dotnet/fsharp/issues/18011

**Comment by ellahathaway** [comment]
Need https://github.com/dotnet/aspnetcore/pull/58987 to unblock aspnetcore signing in this PR.

**Review by ViktorHofer — APPROVED** [review]

**Comment by ellahathaway** [comment]
Interesting. Fsharp is failing on windows with an empty `ItemsToSign` list. Oddly, it seems like signing is being run before the packages are finished being created. That said, it's a bit hard to tell what's occurring, since I only have access to the log and not the binlog. I'll continue to investigate this & try to repro locally.

**Comment by ViktorHofer** [comment]
FYI, https://github.com/dotnet/source-build/issues/4424 is the reason why the AzDO output isn't really helpful here. Unfortunately, fsharp sometimes returns a 0 exit code even though the inner build failed.

**Comment by ellahathaway** [comment]
It looks like building fhsarp on window builds different projects than building fsharp on linux.

Linux builds `FSharp.sln`, whereas Windows builds `VisualFSharp.sln` and `Microsoft.FSharp.Compiler.sln`. From what I can tell, `VisualFSharp.sln` doesn't produce any artifacts, so attempting to sign it fails.

I think in this case, I can condition the fsharp signing to allow an empty ItemsToSignList when doing a DotNetBuild of the `VisualFSharp.sln` file.

@KevinRansom - Do you have a take on this? Is `VisualFSharp.sln` supposed to produce artifacts to sign?

**Comment by ViktorHofer** [comment]
@ellahathaway are you sure about that? When I look at a linux vertical I see the following artifacts getting produced:

```
  New artifact(s) after building fsharp:
    -> FSharp.Compiler.Service/43.9.200-preview.24578.1
    -> FSharp.Core/9.0.200-beta.24578.1
    -> assets/symbols/FSharp.Compiler.Service.43.9.200-preview.24578.1.symbols.nupkg
    -> assets/symbols/FSharp.Core.9.0.200-beta.24578.1.symbols.nupkg
```

File: src/SourceBuild/patches/fsharp/0001-Allow-empty-ItemsToSign-when-building-VisualFSharp-w.patch
Comment (ellahathaway) [reply src/SourceBuild/patches/fsharp/0001-Allow-empty-ItemsToSign-when-building-VisualFSharp-w.patch:24]: Yes, they're just produced later in the build. Fsharp on windows builds two projects, `VisualFSharp.sln` and `Microsoft.FSharp.Compiler.sln`. `VisualFSharp.sln` does not produce any artifacts to sign with core, whereas `Microsoft.FSharp.Compiler.sln` produces the following artifacts with core:

image

**Comment by ellahathaway** [comment]
> @ellahathaway are you sure about that? When I look at a linux vertical (example AzureLinux_x64_Cross_x64) I see the following artifacts getting produced:
> 
> ```
>   New artifact(s) after building fsharp:
>     -> FSharp.Compiler.Service/43.9.200-preview.24578.1
>     -> FSharp.Core/9.0.200-beta.24578.1
>     -> assets/symbols/FSharp.Compiler.Service.43.9.200-preview.24578.1.symbols.nupkg
>     -> assets/symbols/FSharp.Core.9.0.200-beta.24578.1.symbols.nupkg
> ```

Here are the artifacts that I see when I look at the windows vertical:
```
-   New artifact(s) after building fsharp:
    -> Microsoft.FSharp.Compiler/13.9.200-beta.24578.1
    -> assets/symbols/Microsoft.FSharp.Compiler.13.9.200-beta.24578.1.symbols.nupkg
```

Diving into the fsharp binlogs, there are the following items being signed in the linux:
image

By contrast, there are the following items being signed in the windows vertical:
image


It looks like the windows vertical is not considering FSharp.Core to be a shipping artifact, but the artifact is being built.

**Comment by ellahathaway** [comment]
Filed https://github.com/dotnet/fsharp/issues/18112 to address the lack of`ItemsToSign` issue from the fsharp side.

File: src/SourceBuild/content/repo-projects/aspnetcore.proj
Comment (mmitche) [inline src/SourceBuild/content/repo-projects/aspnetcore.proj:8]: Is Sign ever true in source build only mode? Does this need the extra conditional?

**Review by mmitche — APPROVED** [review]
Approved with the question about SB + Signing

**Comment by ViktorHofer** [comment]
> It looks like both FSharp.sln and 'VisualFSharp.slndon't produce artifacts when building with .NET Core MSBuild on windows. Passing-noVisualStudio` does not help the signing issue. That said, we may need to keep it regardless because we are not building with visual studio?

Does that mean that we need to switch to desktop msbuild when building fsharp on Windows?

File: src/SourceBuild/content/repo-projects/roslyn.proj
Comment (ViktorHofer) [inline src/SourceBuild/content/repo-projects/roslyn.proj:10]: Why don't we sign roslyn on non-Windows?

Comment (ellahathaway) [reply src/SourceBuild/content/repo-projects/roslyn.proj:10]: We do: `<!-- Roslyn's build.cmd adds -build automatically. On non-windows, use the default -->`

And from the ubuntu dev log: `/__w/1/vmr/src/roslyn/build.sh --restore --build --pack --publish --sign --ci --configuration Release ...`

File: src/SourceBuild/content/repo-projects/aspnetcore.proj
Comment (ellahathaway) [reply src/SourceBuild/content/repo-projects/aspnetcore.proj:8]: You're right, I don't think it'll ever be true. Given that assumption, I don't have a problem with removing the extra conditional.

File: src/SourceBuild/content/repo-projects/roslyn.proj
Comment (ViktorHofer) [reply src/SourceBuild/content/repo-projects/roslyn.proj:10]: Perfect, thanks

**Comment by ellahathaway** [comment]
> > It looks like both FSharp.sln and 'VisualFSharp.slndon't produce artifacts when building with .NET Core MSBuild on windows. Passing-noVisualStudio` does not help the signing issue. That said, we may need to keep it regardless because we are not building with visual studio?
> 
> Does that mean that we need to switch to desktop msbuild when building fsharp on Windows?

If we want to produce `.vsix` files out of the fsharp build, then yes.

@mmitche - pinging for your thoughts on Viktor's question

File: src/SourceBuild/content/repo-projects/aspnetcore.proj
Comment (ellahathaway) [reply src/SourceBuild/content/repo-projects/aspnetcore.proj]: Removed in https://github.com/dotnet/sdk/pull/44855/commits/b8dc8d637b3a21baf61fec499e31ffd0bac09580

**Pull request merged**
